truss/templates/server/model_wrapper.py

```diff
@@ -166,79 +216,73 @@ async def __call__(
-            # the lock.
-            try:
-                for chunk in response:
-                    response_queue.put(ResponseChunk(chunk))
-                response_queue.put(None)
-            finally:
-                self._predict_lock.release()
-
-        response_generate_thread = Thread(target=queue_response_chunks)
-        response_generate_thread.start()
-        return _response_generator(response_queue)
+                if headers and headers.get("accept") == "application/json":
+                    # In the case of a streaming response, consume stream
+                    # if the http accept header is set, and json is requested.
+                    return await _convert_streamed_response_to_string(async_generator)
+
+                # To ensure that a partial read from a client does not cause the semaphore
+                # to stay claimed, we immediately write all of the data from the stream to a
+                # queue. We then return a new generator that reads from the queue, and then
+                # exit the semaphore block.
+                response_queue: asyncio.Queue = asyncio.Queue()
+                async for chunk in async_generator:
+                    await response_queue.put(ResponseChunk(chunk))
+
+                await response_queue.put(None)
+
+                async def _response_generator():
+                    while True:
+                        chunk = await response_queue.get()
+                        if chunk is None:
+                            return
+                        yield chunk.value
+
+                return _response_generator()
+
+            return processed_response
 
 
 class ResponseChunk:
     def __init__(self, value):
         self.value = value
 
 
-def _response_generator(queue: Queue):
+async def _convert_streamed_response_to_string(response: AsyncGenerator):
+    return "".join([str(chunk) async for chunk in response])
+
+
+def _force_async_generator(gen: Union[Generator, AsyncGenerator]) -> AsyncGenerator:
     """
-    When returning the stream result, simply read from the response queue until a `None`
-    is reached.
+    Takes a generator, and converts it into an async generator if it is not already.
     """
-    while True:
-        chunk = queue.get(timeout=STREAMING_RESPONSE_QUEUE_READ_TIMEOUT_SECS)
-        if chunk is None:
-            return
-        else:
-            yield chunk.value
-
+    if inspect.isasyncgen(gen):
+        return gen
 
-def _convert_streamed_response_to_string(response: Any):
-    return "".join([str(chunk) for chunk in list(response)])
+    async def _convert_generator_to_async():
+        """
+        Runs each iteration of the generator in an offloaded thread, to ensure
+        the main loop is not blocked, and yield to create an async generator.
+        """
+        FINAL_GENERATOR_VALUE = object()
+        while True:
+            chunk = await to_thread.run_sync(next, gen, FINAL_GENERATOR_VALUE)
```

Comment:
A comment here might help, that this is equivalent of running `next(gen, FINAL_GENERATOR_VALUE)` on separate thread. It took me a minute to see this.

Reply:
will add, I agree this is confusing